File: CalenderView/app/src/main/java/com/non_name_hero/calenderview/inputForm/colorSelectActivity.java
package com.non_name_hero.calenderview.inputForm;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.non_name_hero.calenderview.R;
import com.non_name_hero.calenderview.data.ScheduleGroup;
import com.non_name_hero.calenderview.data.source.ScheduleDataSource;
import com.non_name_hero.calenderview.data.source.ScheduleRepository;
import com.non_name_hero.calenderview.utils.Injection;
import com.non_name_hero.calenderview.utils.dialogUtils.PigLeadDeleteDialog;
import com.non_name_hero.calenderview.utils.dialogUtils.PigLeadDialogFragment;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static java.lang.Boolean.FALSE;
import static java.lang.Boolean.TRUE;

public class colorSelectActivity extends AppCompatActivity implements PigLeadDeleteDialog {

    private static final int REQUEST_CODE = 1;
    private static final String DELETE_DIALOG_TAG = "DELETE_DIALOG";

    private Context context;

    private listAdapter listAdapter;
    private ListView listView;

    private Button editButton;
    private Button colorCreateButton;

    private Intent intentOut;
    private ScheduleRepository repository;
    private SharedPreferences prefs;
    private SharedPreferences.Editor editor;

    private boolean createFlag = FALSE;

    //コンストラクタ
    public colorSelectActivity() {

    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.context = this;
        repository = Injection.provideScheduleRepository(context);

        setContentView(R.layout.color_select);
        final Toolbar myToolbar = (Toolbar) findViewById(R.id.colorSelectToolbar);
        editButton = findViewById(R.id.editButton);
        setSupportActionBar(myToolbar);

        //ListViewのIDを取得
        listView = findViewById(R.id.listView);

        //リストのアダプターを使用してViewを作成
        listAdapter = new listAdapter(context, this);
        //削除ダイアログを
        listAdapter.deleteDialog = this;
        listView.setAdapter(listAdapter);

        //DBから情報を取得
        loadColorList();

        //色作成画面用intent
        intentOut = new Intent(this, colorCreateActivity.class);

        colorCreateButton = findViewById(R.id.colorCreateButton);

        /*編集ボタンが押されたとき*************************/
        editButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //ボタンの文字が「編集」ならば
                if (editButton.getText().toString().equals("編集")) {
                    //TODO　リストビューに削除ボタン表示
                    jdgEditMode(TRUE, "完了");
                    //色作成ボタン非表示
                    colorCreateButton.setVisibility(View.GONE);
                }
                //ボタンの文字が「完了ならば」
                else {
                    //TODO　リストビューから削除ボタンを非表示に
                    jdgEditMode(FALSE, "編集");
                    //色作成ボタン表示
                    colorCreateButton.setVisibility(View.VISIBLE);
                }
            }
        });
        /************************************************/

        /*色作成ボタンが押されたとき************************/
        colorCreateButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //色作成画面に遷移
                goColorCreateActivity();
            }
        });
        /************************************************/

        //editFlag判定用Flag
        createFlag = TRUE;

    }

    @Override
    public void onResume() {
        super.onResume();

        //createFlagがTRUEならば
        if (createFlag) {
            //アプリ再開時にeditFlagを0にする
            //TODO　リストビューから削除ボタンを非表示に
            jdgEditMode(FALSE, "編集");
        }
        //それ以外
        else {
            /* 何もしない */
        }

    }

    private void jdgEditMode(boolean value, String str) {
        prefs = getSharedPreferences("input_data", MODE_PRIVATE);
        editor = prefs.edit();
        //SharedPreferenceにeditFlagの値を保存
        editor.putBoolean("editFlag", value);
        //非同期処理ならapply()、同期処理ならcommit()
        editor.commit();
        listView.setAdapter(listAdapter);
        //ボタン文字の切り替え(編集/完了)
        editButton.setText(str);
    }

    public void goColorCreateActivity() {
        //戻り値を設定して色画面に遷移
        startActivityForResult(intentOut, REQUEST_CODE);
    }

    //Activityから戻ってきたときの処理
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        switch (requestCode) {
            //colorCreateActivityから戻ってきた場合
            case (REQUEST_CODE):
                if (resultCode == RESULT_OK) {
                    //DB問い合わせて更新
                    loadColorList();
                } else if (resultCode == RESULT_CANCELED) {
                    //キャンセルボタンを押して戻ってきたときの処理
                } else {
                    //その他
                }
                break;
            default:
                break;
        }
    }

    private void loadColorList() {
        repository.getListScheduleGroup(new ScheduleDataSource.GetScheduleGroupsCallback() {
            @Override
            public void onScheduleGroupsLoaded(List<ScheduleGroup> Groups) {
                //取得後の処理
                listAdapter.setList(Groups);
            }

            @Override
            public void onDataNotAvailable() {

            }
        });
    }

    /**
     * 削除用ダイアログを設定
     *
     * @param callback ダイアログのボタン押下時の処理
     * @return dialog DialogFragmentのオブジェクト
     */
    @Override
    public PigLeadDialogFragment getDeleteDialog(final ScheduleGroup scheduleGroup, @NonNull final DialogCallback callback) {
        // 表示させるメッセージの定義
        final ArrayList<String> dialogMessages = new ArrayList<>(Arrays.asList(getResources().getStringArray(R.array.delete_schedule_group_dialog_massage)));
        // 表示メッセージに削除対象の名前を挿入
        dialogMessages.set(0, String.format(dialogMessages.get(0), scheduleGroup.getGroupName()));
        final String positiveBtnMessage = getString(R.string.delete_schedule_group_positive);
        final String negativeBtnMessage = getString(R.string.delete_schedule_group_negative);
        // AlertDialogを設定
        PigLeadDialogFragment dialog = new PigLeadDialogFragment(context);
        dialog.setDialogMessage(dialogMessages)
                .setPositiveBtnMessage(positiveBtnMessage)
                .setNegativeBtnMessage(negativeBtnMessage)
                .setPositiveClickListener(new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                       callback.onClickPositiveBtn();
                    }
                })
                .setNegativeClickListener(new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        callback.onClickNegativeBtn();
                    }
                });
        return dialog;
    }

    @Override
    public void showPigLeadDiaLog(PigLeadDialogFragment dialog) {
        // AlertDialogを表示
        dialog.show(getSupportFragmentManager(), DELETE_DIALOG_TAG);
    }
}

File: CalenderView/app/src/main/java/com/non_name_hero/calenderview/inputForm/colorSelectActivity.kt
package com.non_name_hero.calenderview.inputForm

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.databinding.DataBindingUtil
import com.non_name_hero.calenderview.R
import com.non_name_hero.calenderview.calendar.CalendarViewModel
import com.non_name_hero.calenderview.calendar.MainActivity
import com.non_name_hero.calenderview.data.ScheduleGroup
import com.non_name_hero.calenderview.data.source.ScheduleDataSource.GetScheduleGroupsCallback
import com.non_name_hero.calenderview.data.source.ScheduleRepository
import com.non_name_hero.calenderview.databinding.ActivityMainBinding
import com.non_name_hero.calenderview.databinding.ColorSelectBinding
import com.non_name_hero.calenderview.utils.Injection
import com.non_name_hero.calenderview.utils.dialogUtils.PigLeadDeleteDialog
import com.non_name_hero.calenderview.utils.dialogUtils.PigLeadDialogBase.DialogCallback
import com.non_name_hero.calenderview.utils.dialogUtils.PigLeadDialogFragment
import com.non_name_hero.calenderview.utils.obtainViewModel

class ColorSelectActivity  /*コンストラクタ*/
    : AppCompatActivity(), PigLeadDeleteDialog {

    private lateinit var binding: ColorSelectBinding                /*ColorSelectActivityのbinding*/

    private lateinit var context: Context                           /*ColorSelectActivityのcontext*/

    private lateinit var ListAdapter: ListAdapter                   /*色グループリストアダプタ*/

    private lateinit var listView: ListView                         /*色グループリストビュー*/

    private lateinit var editButton: Button                         /*色グループ編集ボタン*/
    private lateinit var colorCreateButton: Button                  /*色グループ作成ボタン*/

    private lateinit var repository: ScheduleRepository             /**/


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        /*初期設定***************************************/
        context = this
        repository = Injection.provideScheduleRepository(context)
        //DataBinding
        binding = DataBindingUtil.setContentView(this, R.layout.color_select)
        //LifecycleOwnerを指定
        binding.lifecycleOwner = this
        /*ビューモデル設定*/
        binding.viewModel = obtainViewModel()
        val myToolbar = findViewById<View>(R.id.colorSelectToolbar) as Toolbar
        editButton = findViewById(R.id.editButton)
        colorCreateButton = findViewById(R.id.colorCreateButton)
        setSupportActionBar(myToolbar)

        /*ListViewのIDを取得*/
        listView = findViewById(R.id.listView)

        /*リストのアダプターを使用してViewを作成*/
        ListAdapter = ListAdapter(context, this)
        //削除ダイアログを設定
        ListAdapter.deleteDialog = this
        listView.adapter = ListAdapter

        /*DBから情報を取得*/
        loadColorList()
        /************************************************/

        /*編集ボタンが押されたとき***********************/
        editButton.setOnClickListener {
            /*ボタンの文字が「編集」ならば*/
            if (editButton.text.toString() == "編集") {
                /*リストビューに削除ボタン表示*/
                jdgEditMode(true, "完了")
                /*色作成ボタン非表示*/
                colorCreateButton.visibility = View.GONE
            } else {
                /*リストビューから削除ボタンを非表示に*/
                jdgEditMode(false, "編集")
                /*色作成ボタン表示*/
                colorCreateButton.visibility = View.VISIBLE
            }
        }
        /************************************************/

        /*色作成ボタンが押されたとき********************/
        colorCreateButton.setOnClickListener {
            /*色作成画面に遷移*/
            goColorCreateActivity()
        }
        /************************************************/

    }

//    /*画面表示時処理関数*******************************/
//    public override fun onResume() {
//        super.onResume()
//
//        /*アプリ再開時にeditFlagを0にする*/
//        /*リストビューから削除ボタンを非表示に*/
//        jdgEditMode(false, "編集")
//        /*色作成ボタン表示*/
//        colorCreateButton.visibility = View.VISIBLE
//    }
//
//    /************************************************/

    /*編集モードかを判定する関数********************/
    private fun jdgEditMode(value: Boolean, str: String) {
//        /*SharedPreferenceでeditFlagの値を変更*/
//        val prefs = getSharedPreferences("input_data", MODE_PRIVATE)
//        val editor = prefs.edit()
//        /*SharedPreferenceにeditFlagの値を保存*/
//        editor.putBoolean("editFlag", value)
//        /*非同期処理ならapply()、同期処理ならcommit()*/
//        editor.apply()
        binding.viewModel?.setCurrentEditColorMode(value)
        listView.adapter = ListAdapter
        /*ボタン文字の切り替え(編集/完了)*/
        editButton.text = str
    }

    /************************************************/

    /*色作成画面遷移関数*****************************/
    private fun goColorCreateActivity() {
        /*色作成画面用intent*/
        val intentOut = Intent(this, ColorCreateActivity::class.java)
        /*戻り値を設定して色画面に遷移*/
        startActivityForResult(intentOut, REQUEST_CODE)
    }

    /************************************************/

    /*Activityから戻ってきてからList内容を更新する関数*/
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        when (requestCode) {
            REQUEST_CODE -> when (resultCode) {
                RESULT_OK -> {
                    /*DB問い合わせて更新*/
                    loadColorList()
                }
                RESULT_CANCELED -> {
                    /*キャンセルボタンを押して戻ってきたときの処理*/
                }
                else -> {
                    /*その他*/
                }
            }
            else -> {
            }
        }
    }

    /************************************************/

    /*List内容を更新する関数************************/
    private fun loadColorList() {
        repository.getListScheduleGroup(object : GetScheduleGroupsCallback {
            override fun onScheduleGroupsLoaded(Groups: List<ScheduleGroup>) {
                //取得後の処理
                ListAdapter.setList(Groups)
            }

            override fun onDataNotAvailable() {}
        })
    }
    /************************************************/

    /**
     * 削除用ダイアログを設定
     *
     * @param callback ダイアログのボタン押下時の処理
     * @return dialog DialogFragmentのオブジェクト
     */
    override fun getDeleteDialog(scheduleGroup: ScheduleGroup?, callback: DialogCallback): PigLeadDialogFragment? {
        /*表示させるメッセージの定義*/
        val dialogMessages = ArrayList(listOf(*resources.getStringArray(R.array.delete_schedule_group_dialog_massage)))
        /*表示メッセージに削除対象の名前を挿入*/
        dialogMessages[0] = String.format(dialogMessages[0], scheduleGroup?.groupName)
        val positiveBtnMessage = getString(R.string.delete_schedule_group_positive)
        val negativeBtnMessage = getString(R.string.delete_schedule_group_negative)
        /*AlertDialogを設定*/
        val dialog = PigLeadDialogFragment(context)
        dialog.setDialogMessage(dialogMessages)
                .setPositiveBtnMessage(positiveBtnMessage)
                .setNegativeBtnMessage(negativeBtnMessage)
                .setPositiveClickListener { dialog, which -> callback.onClickPositiveBtn() }
                .setNegativeClickListener { dialog, which -> callback.onClickNegativeBtn() }
        return dialog
    }

    /************************************************/

    /*削除警告ダイアログ表示関数*********************/
    override fun showPigLeadDiaLog(dialog: PigLeadDialogFragment?) {
        /*AlertDialogを表示*/
        dialog?.show(supportFragmentManager, DELETE_DIALOG_TAG)
    }

    /************************************************/

    /*定数定義***************************************/
    companion object {
        private const val REQUEST_CODE = 1
        private const val DELETE_DIALOG_TAG = "DELETE_DIALOG"
    }
    /************************************************/

    /**
     * ViewModelを取得する
     * (this.obtainViewModel(Class: ViewModel)は拡張関数)
     * @return viewModel {CalendarViewModel} カレンダー関連の情報を保持するViewModel
     */
    fun obtainViewModel(): ColorSelectViewModel = this.obtainViewModel(ColorSelectViewModel::class.java)

}